contracts/badger-bridge/BadgerBridgeAdapter.sol

```diff
@@ -96,57 +98,87 @@ contract BadgerBridgeAdapter is OwnableUpgradeable, ReentrancyGuardUpgradeable {
+        bytes32 pHash = keccak256(abi.encode(_token, _slippage, _user, _vault));
         uint256 mintAmount = registry.getGatewayBySymbol("BTC").mint(pHash, _amount, _nHash, _sig);
+
         require(mintAmount > 0, "zero mint amount");
+        
         uint256 fee = _processFee(renBTC, mintAmount, mintFeeBps);
         uint256 mintAmountMinusFee = mintAmount.sub(fee);
-
         uint256 wbtcExchanged;
+
         if (_token == address(wBTC)) {
             // Try and swap and transfer wbtc if token wbtc specified.
             uint256 startBalance = wBTC.balanceOf(address(this));
             if (_swapRenBTCForWBTC(mintAmountMinusFee, _slippage)) {
                 uint256 endBalance = wBTC.balanceOf(address(this));
                 wbtcExchanged = endBalance.sub(startBalance);
-                wBTC.safeTransfer(_destination, wbtcExchanged);
+                
+                if (_vault == address(0)) {
+                    wBTC.safeTransfer(_user, wbtcExchanged);
+                } else {
+                    wBTC.safeIncreaseAllowance(_vault, wbtcExchanged);
+                    ISett(_vault).deposit(wbtcExchanged);
+                    IERC20(_vault).safeTransfer(_user, IERC20(_vault).balanceOf(address(this)));
+                }
+
                 emit Mint(mintAmount, wbtcExchanged, fee);
                 return;
             }
         }
 
         emit Mint(mintAmount, wbtcExchanged, fee);
 
-        renBTC.safeTransfer(_destination, mintAmountMinusFee);
+        if (_vault == address(0)) {
+            renBTC.safeTransfer(_user, mintAmountMinusFee);
+        } else {
+            renBTC.safeIncreaseAllowance(_vault, mintAmountMinusFee);
+            ISett(_vault).deposit(mintAmountMinusFee);
+            IERC20(_vault).safeTransfer(_user, IERC20(_vault).balanceOf(address(this)));
+        }
     }
 
     function burn(
         // user args
         address _token, // either renBTC or wBTC
+        address _vault,
         uint256 _slippage,
         bytes calldata _btcDestination,
         uint256 _amount
     ) external nonReentrant {
-        require(_token == address(renBTC) || _token == address(wBTC), "invalid token address");
+        require(_token == address(renBTC) || _token == address(wBTC) || approvedVaults[_vault], "invalid token address");
+
+        bool isVault = _vault != address(0);
+
+        if (isVault) {
+            IERC20(_vault).transferFrom(msg.sender, address(this), _amount);
+            ISett(_vault).withdrawAll();
+        }
 
         uint256 wbtcTransferred;
         uint256 startBalance = renBTC.balanceOf(address(this));
 
-        if (_token == address(renBTC)) {
+        if (_token == address(renBTC) && !isVault) {
             renBTC.safeTransferFrom(msg.sender, address(this), _amount);
         }
 
         if (_token == address(wBTC)) {
-            wBTC.safeTransferFrom(msg.sender, address(this), _amount);
+            if (!isVault) {
```

Comment:
Let's clean up the nesting a little bit, can prob just do a single `IERC20(_token).safeTransferFrom()` earlier in the fn and get rid of the renbtc check too etc.